src/odemis/acq/align/shift.py

```diff
@@ -19,167 +19,158 @@
-
-        # Locate maximum and map back to original pixel grid
-        ACC = abs(CC)
-        loc1 = ACC.argmax(0)
-        max1 = ACC[(loc1, range(ACC.shape[1]))]
-        loc2 = max1.argmax(0)
-
-        rloc = loc1[loc2]
-        cloc = loc2
+        row_shift = rloc
 
-        rloc -= dft_shift
-        cloc -= dft_shift
-
-        row_shift += rloc / precision
-        col_shift += cloc / precision
-
-        if md2 == 1:
-            row_shift = 0
-        if nd2 == 1:
-            col_shift = 0
+    if cloc > nd2:
+        col_shift = cloc - n
+    else:
+        col_shift = cloc
 
-    return col_shift, row_shift
+    shifts = [row_shift, col_shift]
 
+    if precision == 1:
+        pass
 
-def _UpsampledDFT(data, nor, noc, precision=1, roff=0, coff=0):
-    """
-    Upsampled DFT by matrix multiplies.
-    data (numpy.array): 2d array
-    nor, noc (ints): Number of pixels in the output upsampled DFT, in units
-    of upsampled pixels
-    precision (int): Calculate drift within 1/precision of a pixel
-    roff, coff (ints): Row and column offsets, allow to shift the output array
-                    to a region of interest on the DFT
-    returns (tuple of floats): Drift in pixels
-    """
-    z = 1j  # imaginary unit
-    nr, nc = data.shape
+    else:
+        precision_factor = numpy.array(precision, dtype=float_dtype)
+        shifts = numpy.round(shifts * precision_factor) / precision_factor
+        upsampled_region_size = numpy.ceil(precision_factor * 1.5)
+        # Center of output array at dftshift + 1
+        dftshift = numpy.fix(upsampled_region_size / 2.0)
+        # Matrix multiply DFT around the current shift estimate
+        sample_region_offset = dftshift - shifts * precision_factor
+        cross_correlation = _upsampled_dft(image_product.conj(),
+                                           upsampled_region_size,
+                                           precision_factor,
+                                           sample_region_offset).conj()
+        # Locate maximum and map back to original pixel grid
+        maxima = numpy.unravel_index(numpy.argmax(numpy.abs(cross_correlation)),
+                                     cross_correlation.shape)
+        CCmax = cross_correlation[maxima]
 
-    # Compute kernels and obtain DFT by matrix products
-    kernc = numpy.power(math.e, (-z * 2 * math.pi / (nc * precision)) *
-                                ((fft.ifftshift(arange(0, nc))[:, None]).T - nc // 2) *
-                                (arange(0, noc) - coff)[:, None]
-                       )
+        maxima = numpy.stack(maxima).astype(float_dtype, copy=False)
+        maxima -= dftshift
 
-    kernr = numpy.power(math.e, (-z * 2 * math.pi / (nr * precision)) *
-                                (fft.ifftshift(arange(0, nr))[:, None] - nr // 2) *
-                                ((arange(0, nor)[:, None]).T - roff)
-                       )
+        shifts += maxima / precision_factor
```

Comment:
This whole block about the maximum location should be also correct for the precision == 1, right? You'd just need to have  `precision_factor=1`. Then you could drop the special "locate peak" part of precision == 1, and always use this code.